homeassistant/components/lock/zwave.py

```diff
@@ -197,40 +207,68 @@ def clear_usercode(service):
 
 
 class ZwaveLock(zwave.ZWaveDeviceEntity, LockDevice):
-    """Representation of a Z-Wave switch."""
+    """Representation of a Z-Wave Lock."""
 
     def __init__(self, value):
-        """Initialize the Z-Wave switch device."""
+        """Initialize the Z-Wave lock device."""
         zwave.ZWaveDeviceEntity.__init__(self, value, DOMAIN)
-
         self._node = value.node
         self._state = None
         self._notification = None
         self._lock_status = None
+        self._v2btze = None
+
+        # Enable appropriate workaround flags for our device
+        # Make sure that we have values for the key before converting to int
+        if (value.node.manufacturer_id.strip() and
+                value.node.product_id.strip()):
+            specific_sensor_key = (int(value.node.manufacturer_id, 16),
+                                   int(value.node.product_id, 16))
+            if specific_sensor_key in DEVICE_MAPPINGS:
+                if DEVICE_MAPPINGS[specific_sensor_key] == WORKAROUND_V2BTZE:
+                    self._v2btze = 1
+                    _LOGGER.debug("Polycontrol Danalock v2 BTZE "
+                                  "workaround enabled")
         self.update_properties()
 
     def update_properties(self):
         """Callback on data changes for node values."""
         for value in self._node.get_values(
+                class_id=zwave.const.COMMAND_CLASS_DOOR_LOCK).values():
+            if value.type != zwave.const.TYPE_BOOL:
+                continue
+            if value.genre != zwave.const.GENRE_USER:
+                continue
+            self._state = value.data
+            _LOGGER.debug('Lock state set from Bool value and'
+                          ' is %s', value.data)
+            break
+
+        for value in self._node.get_values(
                 class_id=zwave.const.COMMAND_CLASS_ALARM).values():
             if value.label != "Access Control":
                 continue
             self._notification = LOCK_NOTIFICATION.get(value.data)
-            if self._notification:
-                self._state = LOCK_STATUS.get(value.data)
-                _LOGGER.debug('Lock state set from Access Control value and'
-                              ' is %s', value.data)
+            notification_data = value.data
+            if self._v2btze:
+                for value in (self._node.get_values(
```

Comment:
I see this kind of code all over Z-Wave, what about we introduce a method like this:

(not including other filters already present on [node.get_values](https://github.com/OpenZWave/python-openzwave/blob/c0c8b19f2d6d1224cbdca0a30d5cdbbc80f59ba3/src-api/openzwave/node.py#L417-L447))

```python
get_values(node, class_id=None, index=None, data=None):
    result = []

    kwargs = {}
    if class_id is None:
        kwargs['class_id'] = class_id

    values = node.get_values(**kwargs).values()

    for value in values:
        if index is not None and value.index != index:
            continue

        if data is not None and value.data != data:
            continue

        results.append(value)

    return results
```